src/pretix/base/auth.py

```diff
@@ -160,3 +163,62 @@ def form_authenticate(self, request, form_data):
         u = authenticate(request=request, email=form_data['email'].lower(), password=form_data['password'])
         if u and u.auth_backend == self.identifier:
             return u
+
+
+class NumericAndAlphabeticPasswordValidator:
+
+    def validate(self, password, user=None):
+        has_numeric = any(c in string.digits for c in password)
+        has_alpha = any(c in string.ascii_letters for c in password)
+        if not has_numeric or not has_alpha:
+            raise ValidationError(
+                _(
+                    "Your password must contain both numeric and alphabetic characters.",
+                ),
+                code="password_numeric_and_alphabetic",
+            )
+
+    def get_help_text(self):
+        return _(
+            "Your password must contain both numeric and alphabetic characters.",
+        )
+
+
+class HistoryPasswordValidator:
+
+    def __init__(self, history_length=4):
+        self.history_length = history_length
+
+    def validate(self, password, user=None):
+        from pretix.base.models import User
+
+        if not user or not user.pk or not isinstance(user, User):
+            return
+
+        for hp in user.historic_passwords.order_by("-created")[:self.history_length]:
+            if check_password(password, hp.password):
```

Comment:
Would it be a good idea to hash the password with `make_password` once before and then just compare the hashes instead of recalculating the same hash with `check_password`?

Wouldn’t it even make sense to just check for `exists()` with the precalculated hash as a condition? We make sure with `password_changed()` anyways that only `history_length` number of passwords are saved, so no need for `[:self.history_length]`.

Reply:
> Would it be a good idea to hash the password with make_password once before and then just compare the hashes instead of recalculating the same hash with check_password?

make_password() is not a deterministic hash function, as it generates a random salt for every password, so you can't "compare hashes". We could however use make_password(sha256()) or something, but I'm not really sure if that provides any real advantage?

> Wouldn’t it even make sense to just check for exists() with the precalculated hash as a condition?

No, since make_password() is not deterministic.

> We make sure with password_changed() anyways that only history_length number of passwords are saved, so no need for [:self.history_length].

Yeah, but double-checking never hurts, e.g. in case the history length is changed.